### PR TITLE
Add -notice flag to autogenerate NOTICE files

### PR DESCRIPTION
**Pull request by marclop**
## Description
Adds a -notice flag and a subset of flags tied to -notice. When the flag
is set, generates a `NOTICE` file at the root folder of the project when
used with or stdout when using `-d` for dry runs.

The NOTICE file generation heavily relies on the `go.mod` file and its
contents, Any other dependency managers outside of those reliant on
`go.mod` won't be able to make use of this feature.

By enabling this feature, the binary won't be zery dependency anymore
even though it's still very fast.

The format is fixed unless the Go template is overriden with the flag
`-notice-header`, in which case only the header is overwritten,
maintaining the list of dependencies and licenses within enclosed `=`.

It also changes the default error code from `0` to `255` any time the
error is not nil. Before any unspecified errorcodes defaulted to 0 which
is not the desired behaviour.

## Result

```
$ ./bin/go-licenser -exclude golden -notice -notice-year 2018 -notice-project-name "Go Licenser"

Go Licenser
Copyright 2018-2019 Elasticsearch B.V.

This product includes software developed at Elasticsearch B.V. and
third-party software developed by the licenses listed below.

=========================================================================

gopkg.in/src-d/go-license-detector.v2    Apache-2.0
golang.org/x/exp                         BSD-3-Clause
github.com/sirkon/goproxy                MIT
github.com/hashicorp/go-multierror       MPL-2.0

=========================================================================
```

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -0,0 +1,118 @@
+github.com/alcortesm/tgz v0.0.0-20161220082320-9c5fe88206d7 h1:uSoVVbwJiQipAclBbw+8quDsfcvFjOpI5iCf4p/cqCs=
```
Comment (ppapapetrou76) [inline go.sum]: This looks really huge for 4 direct dependencies.
Maybe a `go mod tidy` could help?

File: error.go [hunk 2]
```diff
@@ -32,8 +32,11 @@ func (e Error) Error() string {
...
 // Code returns the exitcode for the error
 func Code(e error) int {
+	if e == nil {
+		return 0
+	}
 	if err, ok := e.(*Error); ok {
 		return err.code
 	}
-	return 0
+	return 255
```
Comment (ppapapetrou76) [inline error.go]: not sure if I completely understand this change 🤔

File: licensing/notice.go [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+// Validate ensures that the structure is usable by its consumer.
+func (params GenerateNoticeParams) Validate() error {
+	var merr = new(multierror.Error)
+	if params.GoModFile == "" {
+		merr = multierror.Append(merr, errors.New("notice: missing file path"))
+	}
+	if params.AnalyseFunc == nil {
+		merr = multierror.Append(merr, errors.New("notice: AnalyseFunc is missing"))
```
Comment (ppapapetrou76) [inline licensing/notice.go]: `nit` and completely optional. This message is not following the pattern of the other two 😄 🤓

File: main_test.go [hunk 4]
```diff
@@ -104,26 +106,38 @@ func dcopy(src, dest string, info os.FileInfo) error {
 	return nil
 }
 
+var goModAnalyseFunc = genAnalyseFunc([]licensedb.Result{
+	// github.com/sirkon/goproxy
```
Comment (ppapapetrou76) [inline main_test.go]: 👮 commented-out code

File: notice.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+			if fileMod := info.ModTime(); fileMod.Before(startDate) {
+				startDate = fileMod
+			}
+			return nil
+		}); err != nil {
+		return err
+	}
+
+	var noticeWriter = params.out
+	if !params.dry {
```
Comment (ppapapetrou76) [inline notice.go]: Maybe invert the if block `if params.dry` and maybe add a message letting the user know it's running in dry-run mode?

File: notice.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+)
+
+func doNotice(path string, params runParams) error {
+	// TODO: potentially use git to extract the date.
+	var startDate, _ = time.Parse("2006", params.noticeYear)
+	if noticeYear == "" {
+		startDate = time.Now()
+	}
+	if err := filepath.Walk(path,
+		func(p string, info os.FileInfo, err error) error {
```
Comment (ppapapetrou76) [inline notice.go]: Should we move this anon function to a named function. The code looks a bit complex here :(

File: licensing/notice.go [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+	return paths, nil
+}
+
+// getLicenses returns a []Dependency which is populated by the analyseFunc.
+// It relies on the local downloaded dependencies in order for the license
+// discovery mechanism to work. The Dependencies slice is sorted by License.
+func getLicenses(analyseFunc func(args ...string) []licensedb.Result, paths ...string) []Dependency {
+	var dependencies = make([]Dependency, 0, len(paths))
+	for _, res := range analyseFunc(paths...) {
+		arg := strings.Replace(res.Arg, goPkgPath()+"/", "", -1)
```
Comment (ppapapetrou76) [inline licensing/notice.go]: Any chance to get something a bit more readable here?
I really have hard times to follow the code in the next 10-12 lines :(
Maybe adding some comments or extracting pieces to private functions with meaningful names?

**Comment by ppapapetrou76** [comment]
@marclop 
I see a build error `'golint' is not recognized as an internal or external command,
operable program or batch file.`

Why don't we try to install goling like we do in `ecl`, `sotreria` etc.

File: go.sum (hunk 1 above)
Comment (marclop) [reply go.sum]: `go mod tidy` does not help at all with the `go.sum`

File: notice.go (hunk 6 above)
Comment (marclop) [reply notice.go]: I've added a comment

File: licensing/notice.go (hunk 7 above)
Comment (marclop) [reply licensing/notice.go]: I've added comments here

File: main.go [hunk 8]
```diff
@@ -256,3 +317,14 @@ func usageFlag() {
...
 	flag.PrintDefaults()
 }
+
+func openTruncateFile(p string) (*os.File, error) {
+	f, err := os.OpenFile(p, os.O_CREATE|os.O_WRONLY|os.O_APPEND, os.ModePerm)
+	if err != nil {
+		return nil, err
+	}
+
+	f.Truncate(0)
```
Comment (ppapapetrou76) [inline main.go]: again here maybe over-engineering to check for errors, right?

File: notice.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+				}
+				return nil
+			}); err != nil {
+			return err
+		}
+	}
+
+	var noticeWriter = params.out
+	if params.dry {
+		fmt.Fprintf(noticeWriter, "Generating NOTICE...\n\n")
```
Comment (ppapapetrou76) [inline notice.go]: I think I commented before but not sure what happened 😄 
I guess `dry` here means that no files are generated. If so then this message is a bit confusing to the end user. I'd prefer to see something that clearly informs me that no files are generated or that the command is running in dry-run mode.

Comment (marclop) [reply notice.go]: I'll fix that to be outputted no matter what, just changing the message.

**Pull request merged**